src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java

```diff
@@ -1274,4 +1308,179 @@ private int getSizeForFields(List<Field> fields) {
     }
 
     private Map<InstanceKlass, ClassData> classDataCache = new HashMap<>();
+
+    /**
+     * The class implements a buffered output stream for compressed segment data.
+     * It is used inside HeapHprofBinWritter only for heap dump.
+     * Because the current implementation of segment heap dump needs to update
+     * the segment size at segment header, and because it is hard to modify the
+     * compressed data after they are written to file, this class first saves the
+     * uncompressed data into an internal buffer, and then writes through to the
+     * GZIPOutputStream when the whole segment data are ready and the size is updated.
+     * The Internal buffer size will be extended when the segment length exceeds the
+     * current buffer size.
+     * This class defines a switch to turn on/off the segment mode. if turned off,
+     * it behaves same as BufferedOutputStream.
+     * */
+    private class CompressedSegmentOutputStream extends BufferedOutputStream {
```

Comment:
Is there a reason why `CompressedSegmentOutputStream` can't just be called `SegmentOutputStream` and always be enabled? That would simplify a lot of logic that currently is different based on whether or not you are compressing the output.